### ISSUE TITLE
Play include behaves weirdly when conditioned on groups

### ISSUE THREAD

**Issue by yfried** [ansible/ansible#18745]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
Include

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = 
  configured module search path = Default w/o overrides
```

Not Affected:
```
ansible 2.3.0 (devel 8e375913b0) last updated 2016/12/05 12:34:04 (GMT +300)
  lib/ansible/modules/core: (detached HEAD 6890003c4f) last updated 2016/12/04 10:26:59 (GMT +300)
  lib/ansible/modules/extras: (detached HEAD 20ea46642b) last updated 2016/12/04 10:26:59 (GMT +300)
  config file = 
  configured module search path = Default w/o overrides
```
```
ansible 2.1.3.0
  config file = 
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT
Fedora 24
RHEL 7
##### SUMMARY
The following should not be included

```yaml
- include: fileB.yml
  when: "{{ 'localhost' not in groups }}"
```

However, **some** tasks in `fileB.yml` are executed while others are skipped

##### STEPS TO REPRODUCE

fileA.yml
```yaml
- include: fileB.yml
  when: "{{ switch in groups }}"
```

fileB.yml
```yaml
- hosts: localhost
  gather_facts: no
  any_errors_fatal: true
  tasks:
      - name: fail nested
        fail:
          msg:
              a: "this is skipped. groups={{ groups }}"
              b: "{{ 'localhost' not in groups }}"

      - name: debug nested
        debug:
          msg:
              a: "this isn't skipped. groups={{ groups }}"
              b: "{{ 'localhost' not in groups }}"
```

##### EXPECTED RESULTS

```
$ ansible-playbook fileA.yml -vvvv -e switch=localhost

TASK [fail nested] **************************************************************************************************
task path: /tmp/fileB.yml:5
skipping: [localhost] => {
    "changed": false, 
    "skip_reason": "Conditional check failed", 
    "skipped": true
}

TASK [debug nested] *************************************************************************************************
task path: /tmp/fileB.yml:11
skipping: [localhost] => {
    "changed": false, 
    "skip_reason": "Conditional check failed", 
    "skipped": true
}

PLAY RECAP **********************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=0   
```

```
$ ansible-playbook fileA.yml -e switch=all

PLAY [localhost] ****************************************************************************************************

TASK [fail nested] **************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": {"a": "this is skipped. groups={'ungrouped': [u'localhost'], 'all': []}", "b": true}}

NO MORE HOSTS LEFT **************************************************************************************************
	to retry, use: --limit @/tmp/fileA.retry

PLAY RECAP **********************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

```
##### ACTUAL RESULTS
```
$ ansible-playbook -vvvv fileA.yml -e switch=localhost

PLAY [localhost] ***************************************************************

TASK [fail nested] *************************************************************
task path: /tmp/fileB.yml:5
skipping: [localhost] => {
    "changed": false, 
    "skip_reason": "Conditional check failed", 
    "skipped": true
}

TASK [debug nested] ************************************************************
task path: /tmp/fileB.yml:11
ok: [localhost] => {
    "msg": {
        "a": "this isn't skipped. groups={'ungrouped': [u'localhost'], 'all': []}", 
        "b": true
    }
}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0   


```

**Comment by Azenet** [comment]
I'm getting this as well, but it's really weird behavior, depends on the used modules... The file is always included (which I assume is already an issue) but depending on what the task is or where it is included in my playbook it will or won't skip. Rolling back to 2.1 for now.

**Comment by yfried** [comment]
AFAIK, the file is included pre run, and skip is decided at runtime

Regards
Yair

**Comment by jimi-c** [comment]
@yfried-redhat I'm not seeing this behavior on devel, so I'm thinking it may have been fixed a couple of weeks ago when I was working on playbook include bugs:
<pre>
[jamesc@jimi 18745]$ ansible --version
ansible 2.3.0 (devel d3dd11b21c) last updated 2016/12/05 10:40:38 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 6890003c4f) last updated 2016/12/05 10:46:31 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 20ea46642b) last updated 2016/12/05 10:46:31 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/usr/share/ansible']

[jamesc@jimi 18745]$ ansible-playbook -vvv fileA.yml -e switch=localhost
Using /etc/ansible/ansible.cfg as config file

PLAYBOOK: fileA.yml ***********************************************************************************************************************************************************************************************
1 plays in fileA.yml

PLAY [localhost] **************************************************************************************************************************************************************************************************

TASK [fail nested] ************************************************************************************************************************************************************************************************
task path: /home/jamesc/testing/18745/fileB.yml:5
skipping: [localhost] => {
    "changed": false, 
    "skip_reason": "Conditional check failed", 
    "skipped": true
}

TASK [debug nested] ***********************************************************************************************************************************************************************************************
task path: /home/jamesc/testing/18745/fileB.yml:11
skipping: [localhost] => {
    "changed": false, 
    "skip_reason": "Conditional check failed", 
    "skipped": true
}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=0   
</pre>

**Comment by yfried** [comment]
As I've mentioned, it's not in devel. I still thought it needed to be
recorded in case it comes back.

Regards
Yair

**Comment by yfried** [comment]
At @jimi-c request, I've verified that stable_2.2 has the fix.
Could this be ported to 2.1?

**Comment by bcoca** [comment]
dont use moustaches with when:

``` when: "'localhost' not in groups"```

also closing as its already fixed (as mentioned above)

**Comment by ssbarnea** [comment]
@bcoca If this bug is fixed in stable2.2 why is not mentioned in any commit? It quite hard to know what commit was supposed to fix it and to see if the fix was included into a newer release or not. Looking at the bug as fixed in 2.2 would make someone believe that this bug is fixed in 2.2 but as far as I know installing latest release of 2.2 will not fix the issue.

My second question is: is this bug only related to use of quotes? Does this mean that if you do not use quotes with `when` your code would run the same with both 2.1 and 2.1? Thanks.
